controllers/certificates/certificates_controller_test.go

```diff
@@ -0,0 +1,149 @@
+		for _, obj := range tcrds {
+			objs = append(objs, obj)
+		}
+		crdClient = crdclientv1.NewSimpleClientset(objs...)
+	} else {
+		crdClient = crdclientv1.NewSimpleClientset()
+	}
+
+	r := &CertificatesReconciler{CRDClient: crdClient.ApiextensionsV1(), Client: c, Scheme: s, Logger: logger}
+
+	if b.Spec.TargetNamespace != "" {
+		t.Logf("Creating test namespace '%s'", b.Spec.TargetNamespace)
+		t.Run("create-test-namespace", func(t *testing.T) {
+			err := c.Create(
+				context.TODO(),
+				&corev1.Namespace{ObjectMeta: metav1.ObjectMeta{Name: b.Spec.TargetNamespace}},
+				&client.CreateOptions{},
+			)
+			g.Expect(err).To(o.BeNil())
+		})
+	}
+
+	return c, crdClient, r
+}
+
+// testCertificatesReconcilerReconcile simulates the reconciliation process for rolling out and
+// rolling back manifests in the informed target namespace name.
+func testCertificatesReconcilerReconcile(t *testing.T, targetNamespace string) {
+	g := o.NewGomegaWithT(t)
+
+	namespacedName := types.NamespacedName{Namespace: "default", Name: "name"}
+	certificateName := types.NamespacedName{
+		Namespace: targetNamespace,
+		Name:      "shipwright-build-webhook-cert",
+	}
+	req := reconcile.Request{NamespacedName: namespacedName}
+
+	b := &v1alpha1.ShipwrightBuild{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      namespacedName.Name,
+			Namespace: namespacedName.Namespace,
+		},
+		Spec: v1alpha1.ShipwrightBuildSpec{
+			TargetNamespace: targetNamespace,
+		},
+	}
+	crd1 := &crdv1.CustomResourceDefinition{}
+	crd1.Name = "certificates.cert-manager.io"
+	crd2 := &crdv1.CustomResourceDefinition{}
+	crd2.Name = "certmanagers.operator.openshift.io"
+	crds := []*crdv1.CustomResourceDefinition{crd1, crd2}
+	_, _, r := bootstrapCertificatesReconciler(t, b, crds)
+
+	t.Logf("Deploying Certificates Controller against '%s' namespace", targetNamespace)
+	t.Setenv(commonctrl.UseManagedWebhookCerts, "true")
+
+	// rolling out all manifests on the desired namespace, making sure the webhook certificate is created
+	t.Run("rollout-manifests", func(t *testing.T) {
+		ctx := context.TODO()
+		res, err := r.Reconcile(ctx, req)
+		g.Expect(err).To(o.BeNil())
+		g.Expect(res.Requeue).To(o.BeFalse())
+		err = r.Get(ctx, certificateName, &certmanager.Certificate{})
+		g.Expect(err).To(o.BeNil())
+	})
+
+	// rolling back all changes, making sure the webhook certificate is also not found afterwards
+	t.Run("rollback-manifests", func(t *testing.T) {
+		ctx := context.TODO()
+
+		err := r.Get(ctx, namespacedName, b)
+		g.Expect(err).To(o.BeNil())
+
+		// delete the ShipwrightBuild CR triggers the rollback logic so the
+		// reconciliation should remove the objects previously deployed
+		err = r.Delete(ctx, b, &client.DeleteOptions{})
+		g.Expect(err).To(o.BeNil())
+
+		res, err := r.Reconcile(ctx, req)
+		g.Expect(err).To(o.BeNil())
+		g.Expect(res.Requeue).To(o.BeFalse())
+
+		err = r.Get(ctx, certificateName, &certmanager.Certificate{})
+		g.Expect(err).To(o.BeNil())
```

Comment:
You are right, well actually i reverted this because deleting the cr , will give an error on reconcile function, the object is not found, however reverting the test as it is, is given an error, looks like we can nolonger update the deletionTimestamp field. 
`Expected
            <*errors.errorString | 0xc000240500>:
            error: Unable to edit name: metadata.deletionTimestamp field is immutable
            {
                s: "error: Unable to edit name: metadata.deletionTimestamp field is immutable",
            }
        to be nil
`
To be honest i dont see how to fix this